bio/bio.go

```diff
@@ -0,0 +1,307 @@
+/*
+Package bio provides utilities for reading and writing sequence data.
+
+There are many different biological file formats for different applications.
+The poly bio package provides a consistent way to work with each of these file
+formats. The underlying data returned by each parser is as raw as we can return
+while still being easy to use for downstream applications, and should be
+immediately recognizable as the original format.
+*/
+package bio
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"io"
+	"math"
+
+	"github.com/TimothyStiles/poly/bio/fasta"
+	"github.com/TimothyStiles/poly/bio/fastq"
+	"github.com/TimothyStiles/poly/bio/pileup"
+	"github.com/TimothyStiles/poly/bio/slow5"
+	"golang.org/x/sync/errgroup"
+)
+
+// Format is a enum of different parser formats.
+type Format int
+
+const (
+	Fasta Format = iota
+	Fastq
+	Slow5
+	Pileup
+)
+
+// DefaultMaxLineLength variables are defined for performance reasons. While
+// parsing, reading byte-by-byte takes far, far longer than reading many bytes
+// into a buffer. In golang, this buffer in bufio is usually 64kb. However,
+// many files, especially slow5 files, can contain single lines that are much
+// longer. We use the default maxLineLength from bufio unless there is a
+// particular reason to use a different number.
+const defaultMaxLineLength int = bufio.MaxScanTokenSize // 64kB is a magic number often used by the Go stdlib for parsing.
+var DefaultMaxLengths = map[Format]int{
+	Fasta:  defaultMaxLineLength,
+	Fastq:  8 * 1024 * 1024,   // The longest single nanopore sequencing read so far is 4Mb. A 8mb buffer should be large enough for any sequencing.
+	Slow5:  128 * 1024 * 1024, // 128mb is used because slow5 lines can be massive, since a single read can be many millions of base pairs.
+	Pileup: defaultMaxLineLength,
+}
+
+/******************************************************************************
+Aug 30, 2023
+
+Lower level interfaces
+
+******************************************************************************/
+
+// LowLevelParser is a generic interface that all parsers must support. It is
+// very simple, only requiring two functions, Header() and Next(). Header()
+// returns the header of the file if there is one: most files, like fasta,
+// fastq, and pileup do not contain headers, while others like sam and slow5 do
+// have headers. Next() returns a record/read/line from the file format, and
+// terminates on an io.EOF error.
+//
+// Next() may terminate with an io.EOF error with a nil DataType or with a
+// full DataType, depending on where the EOF is in the actual file. A check
+// for this is needed at the last Next(), when it returns an io.EOF error. A
+// pointer is used to represent the difference between a null DataType and an
+// empty DataType.
+type LowLevelParser[DataType fasta.Record | fastq.Read | slow5.Read | pileup.Line, DataTypeHeader fasta.Header | fastq.Header | slow5.Header | pileup.Header] interface {
+	Header() (*DataTypeHeader, error)
```

Comment:
Why does `LowLevelParser` return pointers and `Parser` return values?

Reply:
I think both return pointers. Could you link where `Parser` returns values?

Reply:
Oh my bad I was misreading. Marking resolved.

Reply:
Actually, why are we returning pointers to the structs instead of the structs themselves?